### DISCUSSION TITLE
Error in sim_mutations: "AttributeError: 'generator' object has no attribute 'tables'"

### DISCUSSION THREAD

**Discussion by alanrogers** [tskit-dev/msprime#2029] [Q&A] [answered]
I'm trying to move to the 1.x API for msprime and am getting the following error from sim_mutations:

    Traceback (most recent call last):
      File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/msprime/mutations.py", line 1325, in sim_mutations
        tables = tree_sequence.tables
    AttributeError: 'generator' object has no attribute 'tables'
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "/Users/rogers/Dropbox/tch/archgen/msp/mwe.py", line 108, in <module>
        sim = msp.sim_mutations(treeseq, rate = u_per_site)
      File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/msprime/mutations.py", line 1327, in sim_mutations
        raise ValueError("First argument must be a TreeSequence instance.")
    ValueError: First argument must be a TreeSequence instance.`
    
Would be grateful for suggestions. My code is below:
    
    #!/usr/bin/python3
    
    # Before running this: "python3 pip install msprime"
    import msprime as msp
    import os, sys

    # time parameters in generations
    Txyn = 25920
    Tnd = 15000
    Txy = 3788
    Td = 1734      # age of Denisova fossil
    Ta = 1760      # age of Altai fossil
    Talpha = 1897     # time of Neanderthal admixture

    # population sizes
    Nxyn = 64964.1/2.0  # ancestral population
    Nxy = 44869.2/2.0
    Nn = 9756.8/2.0
    Nx = 20000/2.0 # modern Africa
    Ny = 20000/2.0 # modern Europe

    # admixture
    mAlpha = 0.05
    
    nchromosomes = 1      # number of chromosomes
    basepairs = 2e6   # number of nucleotides per chromosome
    u_per_site = 1.4e-8 # mutation
    
    c = 1e-8 # recombination rate per base pair per generation
    
    # One haploid sample from each of 3 populations: two modern (X,Y),
    # and one archaic (N).
    samples = [
        msp.SampleSet(num_samples=1, population=0, time=0, ploidy=1),  # pop X
        msp.SampleSet(num_samples=1, population=1, time=0, ploidy=1),  # pop Y
        msp.SampleSet(num_samples=1, population=2, time=Ta, ploidy=1), # pop N
    ]
    
    # Population configurations. 
    dem = msp.Demography()
    dem.add_population(
        name = "X",
        description = "African",
        initial_size = Nx
    )
    dem.add_population(
        name = "Y",
        description = "Eurasian",
        initial_size = Ny
    )
    dem.add_population(
        name = "Y2",
        description = "Y before alpha gene flow",
        initial_size = Ny
    )
    dem.add_population(
        name = "N",
        description = "Neanderthal",
        initial_size = Nn
    )
    dem.add_population(
        name = "XY",
        description = "early modern humans",
        initial_size = Nxy
    )
    dem.add_population(
        name = "XYN",
        description = "ancestral population",
        initial_size = Nxyn
    )
    
    # N->Y gene flow
    dem.add_admixture(
        time = Talpha,
        derived = "Y",
        ancestral = ["Y2", "N"],
        proportions = [1-mAlpha, mAlpha]
    )
    
    # X-Y split
    dem.add_population_split(
        time = Txy,
        derived = ["X", "Y2"],
        ancestral = "XY"
    )
    
    # XY-N split
    dem.add_population_split(
        time = Txyn,
        derived = ["XY", "N"],
        ancestral = "XYN"
    )
        
    # Seed for random number generator. Read from /dev/urandom.
    seed = int.from_bytes(os.urandom(4), "big")
    
    # Simulate gene genealogy; return iterator over TreeSequence replicates
    treeseq = msp.sim_ancestry(samples = samples,
                               demography = dem,
                               sequence_length = basepairs,
                               random_seed = seed,
                               recombination_rate = c,
                               num_replicates = nchromosomes
    )
        
    # Put mutations onto the gene genealogy
    sim = msp.sim_mutations(treeseq, rate = u_per_site)

**Comment by alanrogers** [comment, marked as answer]
I think I found an anwer to my own question. When sim_ancestry is called without num_replicates, it returns a value of type `tskit.trees.TreeSequence`:

    >>> ts = msp.sim_ancestry(2, sequence_length=100, random_seed=1234)
    >>> type(ts)
    <class 'tskit.trees.TreeSequence'>

But when it's called with num_replicates, it returns a value of type generator:

    >>> ts = msp.sim_ancestry(2, sequence_length=100, random_seed=1234, num_replicates=2)
    >>> type(ts)
    <class 'generator'>

sim_mutations wants a value of the former type, not a generator.

**Reply by benjeffery** [reply]
Thanks for posting this as it helps others who might hit the same problem! It would be technically possible for `sim_mutations` to be changed to support accepting a sequence of tree sequences as input. Not sure if anyone else thinks that's a good idea, but I assume you're not the only one to have hit this. Alternatively `sim_mutaitons` could provide a more helpful error message when not passed an object that doesn't duck type to a tree sequence.

**Reply by alanrogers** [reply]
If you made `sim_mutations` accept a generator, then the class of its returned value would need to include another layer. You would have genotypes within variants within replicates. And I suspect you would have to include that extra layer even when the argument of `sim_mutations` was just a `TreeSequence`. It might be better to beef up the error message rather than making `sim_mutations` accept a generator.

**Reply by jeromekelleher** [reply]
Thanks @alanrogers all good points.

There is an issue open at #1786. I think it's logical enough if we let sim_mutations accept an iterator, so the canonical example would be
```python
topology_reps = msprime.sim_ancestry(..., num_replicates=100)
mutated_reps =  msprime.sim_mutations(topology_reps, ...)
for ts in mutated_reps:
     # do stuff with ts
```
The nice thing about this is that it's efficient in that nothing really happens until you start running the loop. Also, the same code will now work, regardless of the value of ``num_replicates`` in the first call (which is a good property, I think).

I'd be happy to provide pointers if you'd like to help us out with implementing this!

**Reply by alanrogers** [reply]
If I understand the pseudocode in #1786, there would still be ambiguity about the type of `ts` in

    for ts in mutated_reps:
        # do stuff with ts
     
If `num_replicates = 1` in the call to `sim_ancestry`, `ts` would be a `TreeSequence`. But if `num_replicates = 2`, `ts` would be an iterator across `TreeSequence` objects. This would lead to problems within the "for" loop.

It might be better for `sim_ancestry` to return an iterator in all circumstances, even if there is only 1 replicate, and for `sim_mutations` to fail unless its first argument is an iterator.

**Reply by benjeffery** [reply]
We would break a lot of people's existing code by changing the return of `sim_ancestry`. We avoid this as much as possible! The proposal in #1786 extends the behavior without any breaking changes.

**Reply by jeromekelleher** [reply]
No, if ``num_replicates==1``, ``sim_ancestry`` returns an iterator of length 1, so that works fine. It's only if ``num_replicates=None`` (or not specified) that sim_ancestry returns a single TreeSequence instance. So, yes, you're right, the following would work:
```python
num_replicates = 1
topology_reps = msprime.sim_ancestry(..., num_replicates=num_replicates)
mutated_reps =  msprime.sim_mutations(topology_reps, ...)
for ts in mutated_reps:
     # do stuff with ts
```
But this would break:
```python
num_replicates = None
topology_reps = msprime.sim_ancestry(..., num_replicates=num_replicates)
mutated_reps =  msprime.sim_mutations(topology_reps, ...)
for ts in mutated_reps: # For loop doesn't work.
     # do stuff with ts
```
But I think that's OK, since that's the behaviour you'd already get with sim_ancestry.

Changing the behaviour of ``sim_ancestry`` now is not something we can do, as we've tagged this as 1.0 with the understanding that it's long-term stable.

**Reply by alanrogers** [reply]
Responding to my own previous post: It might be better just to remove the `num_replicates` argument from `sim_ancestry`. The goal of that argument is to spare the user from writing a "for" loop. But if one still needs to loop across replicates when processing the output, then what has been gained? Why not put `sim_ancestry` and `sim_mutations` inside the loop? The API would be simpler, and I think the user's code would be clearer.

**Reply by jeromekelleher** [reply]
"Performance" is the short answer @alanrogers - there's a substantial perf difference when you're doing lots of small simulations. I think that was the motivation in the first place - I agree it's not a great pattern, but the perf difference is compelling.

**Reply by alanrogers** [reply]
I would be happy to work on this in collaboration with @mtisla. Should we just clone the repo, create a branch, and start coding, or are there other steps to take first?

**Reply by benjeffery** [reply]
Our development guide at https://tskit.dev/msprime/docs/stable/development.html has a getting started section. It also links out to [the tskit docs](https://tskit.readthedocs.io/en/latest/development.html#github-workflow) where a suggested github workflow is detailed.

**Reply by jeromekelleher** [reply]
Great, thanks @alanrogers - please do reach out if anything is unclear!
